### PR TITLE
Fix SparkMurmurHash3_32 hash inconsistencies with Apache Spark

### PR DESCRIPTION
**Pull request by jlowe**
#7024 added a Spark variant of Murmur3 hashing, but it is inconsistent with Apache Spark's hash calculations in a few areas:
- `-0.0` and `0.0` are not treated the same by Apache Spark for floats and doubles
- byte and short integral values are upcast to a 32-bit unsigned int (i.e.: zero-filled) before calculating the hash

In addition libcudf allows hashing of timestamp columns but the JNI bindings asserted if timestamp columns were passed in, disabling the ability to hash on timestamps directly.

### CODE REVIEW

File: cpp/include/cudf/detail/utilities/hash_functions.cuh
Comment (ttnghia) [inline cpp/include/cudf/detail/utilities/hash_functions.cuh:633]: Why the arguments are `const&`? I just learn that passing a reference to a device callable function is dangerous---the original variable may reside in host memory.

File: cpp/tests/hashing/hash_test.cpp
Comment (ttnghia) [inline cpp/tests/hashing/hash_test.cpp:368]: Should we write a typed test for those? With procedurally generated data?

File: cpp/include/cudf/detail/utilities/hash_functions.cuh
Comment (jlowe) [reply cpp/include/cudf/detail/utilities/hash_functions.cuh:633]: It's passed by reference because the hash function aliases the data as 32-bit integers to perform the hash on arbitrary types which may be expensive to pass by value.  We want the address of the inputs in memory to avoid that copy.  In this case all of the inputs are coming from values within a `cudf::column` and therefore cannot be in host memory.

File: cpp/tests/hashing/hash_test.cpp
Comment (jlowe) [reply cpp/tests/hashing/hash_test.cpp:368]: > Should we write a typed test for those? With procedurally generated data?

Yes, I think this is a good idea, but arguably applies to all hash functions rather than just this one.  I think that's best done as a followup issue which I'm happy to file.

**Review by ttnghia — APPROVED** [review]

**Review by jrhemstad — APPROVED** [review]

**Comment by jrhemstad** [comment]
@gpucibot merge

File: cpp/tests/hashing/hash_test.cpp
Comment (jlowe) [reply cpp/tests/hashing/hash_test.cpp:368]: Filed #7700 to track this.

**Review by revans2 — COMMENTED** [review]
These changes look fine.  

Is this intended to work with decimal? There are no tests for it, but it is also not explicitly disallowed either. 

It feels like we can add that in as a simple specialization. The raw int value of the DECIMAL32 should be cast to a long and then hashed.  The raw long value of a DECIMAL64 should just be hashed.

**Comment by jlowe** [comment]
> Is this intended to work with decimal? There are no tests for it, but it is also not explicitly disallowed either.

Excellent point.  I'll add support for properly hashing decimal32 and decimal64 along with tests.

**Review by abellina — APPROVED** [review]

File: cpp/include/cudf/detail/utilities/hash_functions.cuh
Comment (mythrocks) [reply cpp/include/cudf/detail/utilities/hash_functions.cuh:634]: > I just learn that passing a reference to a device callable function is dangerous---the original variable may reside in host memory.

@ttnghia, this is only a problem when passing a variable by reference between device and host. In your case, that was a `column_device_view` constructed on `__host__` and passed to a `__device__` function. Note that the underlying pointer could have been copied with impunity.

File: cpp/include/cudf/detail/utilities/hash_functions.cuh
Comment (mythrocks) [inline cpp/include/cudf/detail/utilities/hash_functions.cuh:664]: I won't hold up the PR for this, but there's a trick to reducing the number of specializations, by checking if there's a `.value()` method available on the `key` type. :]

File: java/src/test/java/ai/rapids/cudf/ColumnVectorTest.java
Comment (gerashegalov) [inline java/src/test/java/ai/rapids/cudf/ColumnVectorTest.java:535]: can we have a comment pointing to https://github.com/rapidsai/cudf/pull/7672/files#diff-4eb211c99e029b2b29bc31fb5e890a8c1be89b4d9cbe60f3d4dbc9b33a0042abR281 explaining how expected values were calculated

**Review by mythrocks — APPROVED** [review]

**Review by revans2 — APPROVED** [review]

File: java/src/test/java/ai/rapids/cudf/ColumnVectorTest.java
Comment (jlowe) [reply java/src/test/java/ai/rapids/cudf/ColumnVectorTest.java:535]: I missed this before it merged, but I will be posting a new PR shortly with struct support and can add the comment there.

**Pull request merged**
